File: lib/Rank.js
import lodash from 'lodash'
import {HAND_SIZE} from './Const'
import * as Pattern from './Pattern'

export default class Rank {
  static rank(cards) {
    if (cards.length < HAND_SIZE) return Pattern.MAX
    this.cache = this.cache || new Map([])
    const key = cards.map(b => b.value).join(",")
    let value = this.cache.get(key)
    if (value !== undefined) return value
    value = 0
    if (this.isRoyalFlush(cards)) return value
    value += Pattern.ROYAL_FLUSH
    if (this.isStrightFlush(cards)) return value
    value += Pattern.STRIGHT_FLUSH
    if (this.isFourOfAKind(cards)) return value
    value += Pattern.FOUR_OF_A_KIND
    if (this.isFullHouse(cards)) return value
    value += Pattern.FULL_HOUSE
    if (this.isFlush(cards)) return value
    value += Pattern.FLUSH
    if (this.isStright(cards)) return value
    value += Pattern.STRIGHT
    if (this.isThreeOfAKind(cards)) return value
    value += Pattern.THREE_OF_A_KIND
    if (this.isTwoPair(cards)) return value
    value += Pattern.TWO_PAIR
    if (this.isPair(cards)) return value
    value += Pattern.PAIR
    this.cache.set(key, value)
    return value
  }
  static isRoyalFlush(cards) {
    return this.isStrightAce(cards) && this.isFlush(cards)
  }
  static isStrightFlush(cards) {
    return this.isFlush(cards) && this.isStright(cards)
  }
  static isFourOfAKind(cards) {
    return Object.values(this.pairs(cards)).map(a => a.length).some(a => a === 4)
  }
  static isFullHouse(cards) {
    return Object.keys(this.pairs(cards)).length === 2
  }
  static isFlush(cards) {
    return cards.every(a => a.suit === cards[0].suit)
  }
  static isStright(cards) {
    const ranks = cards.map(a => a.rank).sort((a, b) => a - b)
    if (lodash.isEqual([0, 9, 10, 11, 12], ranks)) return true
    return lodash.every(ranks, (v, i) => v === ranks[0] + i) || this.isStrightAce(cards)
  }
  static isStrightAce(cards) {
    const ranks = cards.map(a => a.rank).sort((a, b) => a - b)
    return lodash.isEqual([0, 9, 10, 11, 12], ranks)
  }
  static isThreeOfAKind(cards) {
    return Object.values(this.pairs(cards)).map(a => a.length).some(a => a === 3)
  }
  static isTwoPair(cards) {
    return Object.keys(this.pairs(cards)).length === 3
  }
  static isPair(cards) {
    return Object.keys(this.pairs(cards)).length === 4
  }
  static isHighCard(cards) {
    return Object.keys(this.pairs(cards)).length === 5
  }
  static pairs(cards) {
    return lodash.groupBy(cards, a => a.rank)
  }
}

File: lib/rank.go
package lib

import (
	"fmt"
	"sort"
)

// SortedCards ...
type SortedCards Cards

// Rank ...
func (cards Cards) Rank() Hand {
	hand := Hand{}
	if len(cards) != 5 {
		return hand
	}
	hand.Cards = cards.Copy().Sort()
	hand.Pairs = hand.Cards.Pairs()
	hand.isFlush = hand.IsFlush()
	hand.isStraight = hand.IsStraight()
	hand.highCard = hand.Cards[0]
	if !hand.isFlush && !hand.isStraight && len(hand.Pairs[0]) == 1 {
		hand.rank = HighCard
	} else if !hand.isFlush && len(hand.Pairs[0]) == 2 && len(hand.Pairs[1]) == 2 {
		hand.rank = TwoPair
	} else if !hand.isFlush && len(hand.Pairs[0]) == 2 {
		hand.rank = Pair
	} else if len(hand.Pairs[0]) == 3 && len(hand.Pairs[1]) == 2 {
		hand.rank = FullHouse
	} else if !hand.isFlush && len(hand.Pairs[0]) == 3 {
		hand.rank = ThreeOfAKind
	} else if len(hand.Pairs[0]) == 4 {
		hand.rank = FourOfAKind
	} else if hand.isStraight && hand.isFlush && hand.Cards[0].Rank == ACE && hand.Cards[1].Rank == KING {
		hand.rank = RoyalFlush
	} else if hand.isStraight && hand.isFlush {
		hand.rank = StrightFlush
	} else if hand.isFlush {
		hand.rank = Flush
	} else if hand.isStraight {
		hand.rank = Stright
	}
	return hand
}

// IsFlush ...
func (hand Hand) IsFlush() bool {
	for i := range hand.Cards {
		if hand.Cards[0].Suit != hand.Cards[i].Suit {
			return false
		}
	}
	return true
}

// IsStraight ...
func (hand Hand) IsStraight() bool {
	for i := range hand.Cards {
		a, b := hand.Cards[0], hand.Cards[i]
		if i == 1 && a.Rank == ACE && b.Rank == KING {
			continue
		} else if a.Rank != b.Rank+Rank(i) {
			return false
		}
	}
	return true
}

// Kicker ...
func (hand Hand) Kicker() *Card {
	return hand.Cards[4] //TODO
}

// ToString ...
func (hand Hand) ToString() string {
	return fmt.Sprintf(
		"hand:%s\trank:%d\tname:%s",
		Cards(hand.Cards).ToString(),
		hand.rank,
		HandNames[hand.rank])
}

// Sort ...
func (cards Cards) Sort() SortedCards {
	sort.Slice(cards, func(i, j int) bool {
		if cards[i].Rank != cards[j].Rank {
			if cards[i].Rank == ACE {
				return true
			} else if cards[j].Rank == ACE {
				return false
			}
			return cards[i].Rank > cards[j].Rank
		}
		return cards[i].Suit > cards[j].Suit
	})
	return SortedCards(cards)
}

// Pairs ...
func (cards SortedCards) Pairs() []Cards {
	pairs := make([]Cards, 0)
	for i := range cards {
		flag := true
		for j := range pairs {
			if pairs[j][0].Rank == cards[i].Rank {
				pairs[j] = append(pairs[j], cards[i])
				flag = false
				break
			}
		}
		if flag {
			pairs = append(pairs, Cards{cards[i]})
		}
	}
	sort.Slice(pairs, func(i, j int) bool {
		return len(pairs[i]) > len(pairs[j])
	})
	return pairs
}
